corehq/apps/users/models.py

```diff
@@ -3107,18 +3109,37 @@ class Meta(object):
         unique_together = ('user', 'name')
 
     def save(self, *args, **kwargs):
-        if not self.key:
-            self.key = self.generate_key()
+        if not self.plaintext_key:
+            self.plaintext_key = self._generate_key()
             if 'update_fields' in kwargs:
                 kwargs['update_fields'].append('key')
+                kwargs['update_fields'].append('encrypted_key')
 
         return super().save(*args, **kwargs)
 
-    def generate_key(self):
+    def _generate_key(self):
         # From tastypie
         new_uuid = uuid4()
         return hmac.new(new_uuid.bytes, digestmod=sha1).hexdigest()
 
+    @property
+    def plaintext_key(self):
+        try:
+            decrypted_key = b64_aes_cbc_decrypt(self.encrypted_key)
+            if decrypted_key == self.key:
+                return decrypted_key
+            else:
+                logging.warning("Decrypted key does not match stored key for %s", self.name)
+                return self.key
+        except Exception as e:
+            logging.exception(f'Error getting decrypted key for {self.name}. {e}')
+            return self.key
+
+    @plaintext_key.setter
+    def plaintext_key(self, plaintext):
+        self.key = plaintext
```

Comment:
I don't think we should ever store the key in plaintext, because we don't need to. We can store it encrypted, with a prefix, and the prefix will tell us everything we need to know to decrypt it.

Reply:
This PR is a "transitionary" change where we will write the API key as both plaintext and encrypted. There will be a followup PR that will delete the plaintext "key" field so that we will read only from the "encrypted_key" field.